stacks-signer/src/chainstate.rs

```diff
@@ -0,0 +1,375 @@
+                "proposed_block_pubkey_hash" => %block_pkh,
+                "sortition_winner_pubkey_hash" => %proposed_by.state().miner_pkh,
+            );
+            return Ok(false);
+        }
+
+        // check that this miner is the most recent sortition
+        match proposed_by {
+            ProposedBy::CurrentSortition(sortition) => {
+                if sortition.invalidated != SortitionMinerStatus::Valid {
+                    warn!(
+                        "Current miner behaved improperly, this signer views the miner as invalid.";
+                        "proposed_block_consensus_hash" => %block.header.consensus_hash,
+                        "proposed_block_signer_sighash" => %block.header.signer_signature_hash(),
+                    );
+                    return Ok(false);
+                }
+            }
+            ProposedBy::LastSortition(_last_sortition) => {
+                if let Some(cur_sortition) = &self.cur_sortition {
+                    // should only consider blocks from the last sortition if the new sortition was invalidated
+                    //  before we signed their first block.
+                    if cur_sortition.invalidated
+                        != SortitionMinerStatus::InvalidatedBeforeFirstBlock
+                    {
+                        warn!(
+                            "Miner block proposal is from last sortition winner, when the new sortition winner is still valid. Considering proposal invalid.";
+                            "proposed_block_consensus_hash" => %block.header.consensus_hash,
+                            "proposed_block_signer_sighash" => %block.header.signer_signature_hash(),
+                        );
+                        return Ok(false);
+                    }
+                }
+            }
+        };
+
+        if let Some(tenure_change) = block.get_tenure_change_tx_payload() {
+            // in tenure changes, we need to check:
+            // (1) if the tenure change confirms the expected parent block (i.e.,
+            //     the last block we signed in the parent tenure)
+            // (2) if the parent tenure was a valid choice
+            let confirms_expected_parent =
+                Self::check_tenure_change_block_confirmation(tenure_change, block, signer_db)?;
+            if !confirms_expected_parent {
+                return Ok(false);
+            }
+            // now, we have to check if the parent tenure was a valid choice.
+            let is_valid_parent_tenure =
+                Self::check_parent_tenure_choice(proposed_by.state(), block, client)?;
+            if !is_valid_parent_tenure {
+                return Ok(false);
+            }
+            let last_in_tenure = signer_db
+                .get_last_signed_block_in_tenure(&block.header.consensus_hash)
+                .map_err(|e| ClientError::InvalidResponse(e.to_string()))?;
+            if last_in_tenure.is_some() {
```

Comment:
I don't think this is right.  What happens if the data loaded from `signer_db` no longer corresponds to canonical Bitcoin state?  Would that cause the signer to erroneously treat this (forked but now canonical) Stacks state as invalid, and prefer the (now invalid) Stacks state written to the `signer_db`?

Reply:
The data from the signerdb is addressed/looked up by the tenure’s consensus hash. Even in the presence of bitcoin forks, those blocks are still valid with respect to consensus hashes. It may be the case that we want to track canonical status in the signerdb for some other reason, but this struct at least tries to be agnostic to this.

Reply:
Yes, I see that now.

No need to track which blocks are canonical -- they're all valid with respect to their Bitcoin forks.